File: app/src/main/java/com/guzzardo/android/willyshmo/kotlintictacdoh/PlayersOnlineActivity.java
package com.guzzardo.android.willyshmo.kotlintictacdoh;

import android.app.Activity;
import android.app.FragmentManager;
import android.app.FragmentTransaction;
import android.app.ListFragment;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.Resources;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Toast;

import com.guzzardo.android.willyshmo.kotlintictacdoh.MainActivity.UserPreferences;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.UnsupportedEncodingException;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Set;
import java.util.TreeMap;

public class PlayersOnlineActivity extends Activity implements ToastMessage {
	
    private String mUsersOnline;
    private static String [] mUserNames;
    private static String [] mUserIds;  
    private static Integer mPlayer1Id;
    private static String mPlayer1Name;
    private static Context mApplicationContext;
    public static ErrorHandler errorHandler;   
    private static Resources mResources;
    private static PlayersOnlineActivity mPlayersOnlineActivity;
    private static int mSelectedPosition = -1;
    private static RabbitMQMessageConsumer mMessageConsumer;
    private static RabbitMQPlayerResponseHandler mRabbitMQPlayerResponseHandler;
    private String mRabbitMQPlayerResponse;

    //TODO - consider saving mUserNames and mUserIds in savedInstanceState and changing AndroidManifest.PlayersOnlineActivity 
    // android:noHistory to false so that we can restore prior list when user presses back button in GameActivity
    // This will make replaying simpler since the user will see the prior list of users online when leaving a game instead of 
    // being sent back to the 2 Players screen.
    // The downside is that this list will become increasingly inaccurate as other players enter and leave the online list.
    
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        
        mResources = getResources();
        errorHandler = new ErrorHandler();
        mApplicationContext = getApplicationContext();
        getSharedPreferences();
        mPlayersOnlineActivity = this;
        mPlayer1Name = getIntent().getStringExtra(GameActivity.PLAYER1_NAME);
        getPlayersOnline();
        resetPlayersOnline();
        	
        if (mUserNames.length == 0) {
        	writeToLog("PlayersOnlineActivity", "starting server only");
           	Intent i = new Intent(mApplicationContext, GameActivity.class);
           	i.putExtra(GameActivity.START_SERVER, "true");
           	i.putExtra(GameActivity.PLAYER1_ID, mPlayer1Id); 
           	i.putExtra(GameActivity.PLAYER1_NAME, mPlayer1Name);
           	startActivity(i);
           	finish();
        } else {
            mMessageConsumer = new RabbitMQMessageConsumer(this, mResources);
            mRabbitMQPlayerResponseHandler = new RabbitMQPlayerResponseHandler();
            setContentView(R.layout.players_online); //this starts up the list view
        }
    }
    
    private class RabbitMQPlayerResponseHandler implements RabbitMQResponseHandler {
    	public void setRabbitMQResponse(String rabbitMQResponse) {
  	  		mRabbitMQPlayerResponse = rabbitMQResponse;
  	  	}
  	  	public String getRabbitMQResponse() {
  	  		return mRabbitMQPlayerResponse;
  	  	}
    }
    
  	private static void setUpMessageConsumer(RabbitMQMessageConsumer rabbitMQMessageConsumer, final String qNameQualifier, final RabbitMQResponseHandler rabbitMQResponseHandler) { 
  		//String hostName = mResources.getString(R.string.RabbitMQHostName);
        String hostName = (String)WillyShmoApplication.getConfigMap("RabbitMQIpAddress");
        String queuePrefix = (String)WillyShmoApplication.getConfigMap("RabbitMQQueuePrefix");
  		String qName = queuePrefix + "-" + qNameQualifier + "-" + mPlayer1Id;
  		new ConsumerConnectTask().execute(hostName, rabbitMQMessageConsumer, qName, mPlayersOnlineActivity, mResources, "fromPlayersOnlineActivity");
  		writeToLog("PlayersOnlineActivity", qNameQualifier +" message consumer listening on queue: " + qName);		
		
  		// register for messages
  		rabbitMQMessageConsumer.setOnReceiveMessageHandler(new RabbitMQMessageConsumer.OnReceiveMessageHandler() {
  			public void onReceiveMessage(byte[] message) {
  				String text = "";
  				try {
  					text = new String(message, "UTF8");
  				} catch (UnsupportedEncodingException e) {
  					mPlayersOnlineActivity.sendToastMessage(e.getMessage());
  				}
                rabbitMQResponseHandler.setRabbitMQResponse(text);
  				writeToLog("PlayersOnlineActivity", qNameQualifier + " OnReceiveMessageHandler received message: " + text);	
  			}
  		});
  	}

    @Override
    public void onPause() {
    	super.onPause();
    	if (mSelectedPosition == -1) {
    		String urlData = "/gamePlayer/update/?id=" + mPlayer1Id + "&onlineNow=false&opponentId=0&userName=";
    		new SendMessageToWillyShmoServer().execute(urlData, mPlayer1Name, this, mResources, false);
    	}
    }
    
    /**
     * This is the "top-level" fragment, showing a list of items that the
     * user can pick.  Upon picking an item, it takes care of displaying the
     * data to the user as appropriate based on the currrent UI layout.
     */

    public static class PlayersOnlineFragment extends ListFragment {
        private boolean mDualPane;
        private int mCurCheckPosition = 0;

        @Override
        public void onActivityCreated(Bundle savedInstanceState) {
            super.onActivityCreated(savedInstanceState);

            // Populate list with our static array of titles.
            setListAdapter(new ArrayAdapter<String>(getActivity(),
                    android.R.layout.simple_list_item_activated_1, mUserNames));

            // Check to see if we have a frame in which to embed the details
            // fragment directly in the containing UI.
            //View detailsFrame = getActivity().findViewById(R.id.details);
            //mDualPane = detailsFrame != null && detailsFrame.getVisibility() == View.VISIBLE;

            if (savedInstanceState != null) {
                // Restore last state for checked position.
                mCurCheckPosition = savedInstanceState.getInt("curChoice", 0);
            }

            if (mDualPane) {
                // In dual-pane mode, the list view highlights the selected item.
                getListView().setChoiceMode(ListView.CHOICE_MODE_SINGLE);
                // Make sure our UI is in the correct state.
                showDetails(mCurCheckPosition);
            }
       }
        
        private void showAcceptDialog(String opponentName, String opponentId) {
            FragmentManager manager = this.getFragmentManager();
            FragmentTransaction ft = manager.beginTransaction();

            Bundle myBundle = new Bundle();
            myBundle.putString("opponentName", opponentName);
            myBundle.putString("opponentId", opponentId);
            myBundle.putString("playerName", mPlayer1Name);
            myBundle.putInt("player1Id", mPlayer1Id);
            AcceptGameDialog acceptGameDialog = new AcceptGameDialog();
            //AcceptGameDialog acceptGameDialog = new AcceptGameDialog(opponentName, opponentId, mPlayer1Name, mPlayer1Id, mApplicationContext, mResources);
            acceptGameDialog.setArguments(myBundle);
            acceptGameDialog.setContext(mApplicationContext);
            acceptGameDialog.setResources(mResources);
            acceptGameDialog.show(ft, "dialog");
        }
        
        private void showRejectDialog(String opponentName, String opponentId) {
            FragmentManager manager = this.getFragmentManager();
            FragmentTransaction ft = manager.beginTransaction();

            Bundle myBundle = new Bundle();
            myBundle.putString("opponentName", opponentName);
            myBundle.putString("opponentId", opponentId);
            myBundle.putString("playerName", mPlayer1Name);
            myBundle.putInt("player1Id", mPlayer1Id);
            RejectGameDialog rejectGameDialog = new RejectGameDialog();
            rejectGameDialog.setArguments(myBundle);
            rejectGameDialog.setContext(mApplicationContext);
            rejectGameDialog.setResources(mResources);
            rejectGameDialog.show(ft, "dialog");
        }
        
        @Override
        public void onResume() { //only called when at least one opponent is online to select 
            super.onResume();
            startGame();

//          This doesnt seen to work, the intent is, if there is only one selection available, then select it automatically
//            if (mUserNames.length == 1)
//                getListView().setSelection(0);


//         	mWaitForOpponent = new ListThread();
//         	mWaitForOpponent.start();
        }

        @Override
        public void onPause() {
        	super.onPause();
        	//mWaitingForOpponent = false;
         	new DisposeRabbitMQTask().execute(mMessageConsumer, mResources, mPlayersOnlineActivity);
        }

        @Override
        public void onSaveInstanceState(Bundle outState) {
            super.onSaveInstanceState(outState);
            outState.putInt("curChoice", mCurCheckPosition);
        }

        @Override
        public void onListItemClick(ListView l, View v, int position, long id) {
            setUpClientAndServer(position);

            String hostName = (String)WillyShmoApplication.getConfigMap("RabbitMQIpAddress");
            String queuePrefix = (String)WillyShmoApplication.getConfigMap("RabbitMQQueuePrefix");
   			String qName = queuePrefix + "-" + "startGame"  + "-" +  mUserIds[position];
   			
   			String messageToOpponent = "letsPlay," + mPlayer1Name + ","  + mPlayer1Id; //mUserIds[position];
   			new SendMessageToRabbitMQTask().execute(hostName, qName, null, messageToOpponent, mPlayersOnlineActivity, mResources);  
            mSelectedPosition = position;
        }
        
        private void setUpClientAndServer(int which) {
	        SharedPreferences settings = mApplicationContext.getSharedPreferences(UserPreferences.PREFS_NAME, 0);
	        SharedPreferences.Editor editor = settings.edit();
	        editor.putString("ga_opponent_screenName", mUserNames[which]);
            editor.apply();
        	
        	Intent i = new Intent(mApplicationContext, GameActivity.class);
        	i.putExtra(GameActivity.START_SERVER, "true");
            i.putExtra(GameActivity.START_CLIENT, "true"); //this will send the new game to the client
        	i.putExtra(GameActivity.PLAYER1_ID, mPlayer1Id); 
        	i.putExtra(GameActivity.PLAYER1_NAME, mPlayer1Name);
        	i.putExtra(GameActivity.START_CLIENT_OPPONENT_ID, mUserIds[which]);
            i.putExtra(GameActivity.PLAYER2_NAME, mUserNames[which]);
            i.putExtra(GameActivity.START_FROM_PLAYER_LIST, "true"); 
            writeToLog("PlayersOnlineActivity", "starting client and server");
        	startActivity(i);
        }

        void showDetails(int index) {
            mCurCheckPosition = index;
        }
    }

    static private void startGame() {
        setUpMessageConsumer(mMessageConsumer, "startGame", mRabbitMQPlayerResponseHandler);
        mSelectedPosition = -1;
        mRabbitMQPlayerResponseHandler.setRabbitMQResponse("null"); // get rid of any old game requests
    }
    
    private void getPlayersOnline() {
    	TreeMap<String, HashMap<String, String>> users = parseUserList(mUsersOnline);

    	TreeMap<String, HashMap<String, String>> usersClone = (TreeMap<String, HashMap<String, String>>) users.clone();    	
    	//we're creating a clone because removing an entry from the original TreeMap causes a problem for the iterator
    	Set<String> userKeySet = usersClone.keySet(); // this is where the keys (userNames) gets sorted
    	Iterator<String> keySetIterator = userKeySet.iterator();

    	while (keySetIterator.hasNext()) {
    		String key = keySetIterator.next();
    		HashMap<String, String> userValues = users.get(key);
    		String userId = userValues.get("userId");
    		if (userId.equals(Integer.toString(mPlayer1Id))) //not going to play against myself on the network
    			users.remove(key);
    	}
    	
    	Object[] objectArray = users.keySet().toArray();
    	mUserNames = new String[objectArray.length];
    	mUserIds = new String[objectArray.length];
    	for (int x = 0; x < objectArray.length; x++) {
    		mUserNames[x] = (String)objectArray[x];
    		
            HashMap<String, String> userValues = users.get(objectArray[x]);
            mUserIds[x] =  userValues.get("userId");
            //System.out.println("user id: " + mUserIds[x] + " for: " + mUserNames[x]);
    	}
    }
    
    private TreeMap<String, HashMap<String, String>> parseUserList(String usersLoggedOn) {
    	TreeMap<String, HashMap<String, String>> userTreeMap = new TreeMap<String, HashMap<String, String>>();
    	
    	try {
    		JSONObject jsonObject = new JSONObject(usersLoggedOn);
    		JSONArray userArray = jsonObject.getJSONArray("UserList"); 
        	
        	for (int y = 0; y < userArray.length(); y++) {
        		HashMap<String, String> userMapValues = new HashMap<String, String>();
        		JSONObject userValues = userArray.getJSONObject(y);
        		String userId = userValues.getString("id");
        		String userName = userValues.getString("userName");
        		userMapValues.put("userId", userId);
        		userTreeMap.put(userName, userMapValues);  
        	}
        } catch (JSONException e) {
         	writeToLog("PlayersOnlineActivity", "parseUserList: " + e.getMessage());
			sendToastMessage(e.getMessage());			
        }
        return userTreeMap;
    }
    
    private void getSharedPreferences() {
        SharedPreferences settings = getSharedPreferences(UserPreferences.PREFS_NAME, MODE_PRIVATE);
        mUsersOnline = settings.getString("ga_users_online", null);
        mPlayer1Id = settings.getInt(GameActivity.PLAYER1_ID, 0); 
    }
    
    private void resetPlayersOnline() {
        SharedPreferences settings = getSharedPreferences(UserPreferences.PREFS_NAME, 0);
        SharedPreferences.Editor editor = settings.edit();
        editor.putString("ga_users_online", null);
        editor.apply();
    }

    /**
     * A simple utility Handler to display an error message as a Toast popup
     */
    
    private class ErrorHandler extends Handler {
        @Override
        public void handleMessage(Message msg) {
    		Toast.makeText(getApplicationContext(), (String)msg.obj, Toast.LENGTH_LONG).show();
        }
    }
    
    public void sendToastMessage(String message) {
    	Message msg = PlayersOnlineActivity.errorHandler.obtainMessage();
    	msg.obj = message;
    	PlayersOnlineActivity.errorHandler.sendMessage(msg);	
    }
    
    private static void writeToLog(String filter, String msg) {
    	if ("true".equalsIgnoreCase(mResources.getString(R.string.debug))) {
    		Log.d(filter, msg);
    	}
    }
    
}

File: app/src/main/java/com/guzzardo/android/willyshmo/kotlintictacdoh/PlayersOnlineActivity.kt
package com.guzzardo.android.willyshmo.kotlintictacdoh

import android.content.Context
import android.content.Intent
import android.content.res.Resources
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ListView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.core.content.ContextCompat.startActivity
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.ListFragment
import com.guzzardo.android.willyshmo.kotlintictacdoh.MainActivity.UserPreferences
import com.guzzardo.android.willyshmo.kotlintictacdoh.WillyShmoApplication.Companion.getConfigMap
import kotlinx.coroutines.*
import org.json.JSONException
import org.json.JSONObject
import java.nio.charset.StandardCharsets
import java.text.SimpleDateFormat
import java.util.*

/*
We can start GameActivity from PlayersOnlineActivity 3 different ways:

1. In onCreate, if there are no other users online, starts server only - Cannabis.
   In this scenario, GameActivity will listen on queue test-client-xx.

2. When an opposing player is selected from the Fragment List, starts client and server - Dr. Strangelove.
   In this scenario PlayersOnlineActivity will send a letsPlay message to the opposing player's PlayersOnlineActivity.

3. When a letsPlay message is received from an opposing player, starts server only - Shakespeare.

Only PlayersOnlineActivity is primed to accept messages using the qName prefix of test-playerList-XX where XX is the PlayerID of the opposing player.

Note - letsPlay message is sent from one place only, PlayersOnlineFragment and is accepted only by PlayersOnlineActivity.
leftGame only comes from onPause in GameActivity.

Willy Shmo play over web messaging sequence:

1. Player selects name from list of users online - client sends letsPlay message to server from PlayersOnlineActivity
2. Server picks up this message in PlayersOnlineActivity
3. client sends tokenList to server:
   PlayersOnlineActivity sends intent to GameActivity: intent.putExtra(GameActivity.START_CLIENT, "true")
   This will send the new game to the client in a tokenList message.

If no one else is online, then starting a new game will start the server side, display the game board and wait for a tokenList message from the opponent.

D/SendMessageToRabbitMQTask: message: letsPlay,Evelyn,81, 738, 2021-08-08 22:54:22 to queue: test-startGame-90
D/GameActivity: client message consumer listening on queue: test-client-81
D/PlayersOnlineActivity: startGame message consumer listening on queue: test-startGame-81

Message sequence:
D/GameActivity: server OnReceiveMessageHandler received message: {"tokenList"... - received from opposing player
D/ServerThread: Server responded to client completed, queue: test-client-202, message: serverAccepted
GameActivity commences game play
*/

class PlayersOnlineActivity : FragmentActivity(), ToastMessage {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mSavedInstanceState = savedInstanceState
        mResources = resources
        mErrorHandler = ErrorHandler()
        mApplicationContext = applicationContext
        mCallerActivity = this
        sharedPreferences
        mPlayersOnlineActivity = this
        mPlayer1Name = intent.getStringExtra(GameActivity.PLAYER1_NAME)

        if (mUsersOnline.isNullOrEmpty()) {
            writeToLog("PlayersOnlineActivity", "onCreate() mUsersOnline == null")
            sendToastMessage("Sorry unable to retrieve Users online, please try again")
            finish()
        } else {
            playersOnline
            setNumberOfPlayersOnline()
            if (mUserNames.isEmpty()) {
                writeToLog("PlayersOnlineActivity", "starting server only")
                val intent = Intent(mApplicationContext, GameActivity::class.java)
                val item = GameActivity.ParcelItems(123456789, "Cannabis")
                intent.putExtra(GameActivity.PARCELABLE_VALUES, item)
                intent.putExtra(GameActivity.START_SERVER, "true")
                intent.putExtra(GameActivity.PLAYER1_ID, mPlayer1Id)
                intent.putExtra(GameActivity.PLAYER1_NAME, mPlayer1Name)
                startActivity(intent)
                finish()
            } else {
                mMessageConsumer = RabbitMQMessageConsumer(this, mResources)
                mMessageConsumer!!.setUpMessageConsumer("playerList", mPlayer1Id, this, mResources, "PlayersOnlineActivity")
                mMessageConsumer!!.setOnReceiveMessageHandler(object :
                    RabbitMQMessageConsumer.OnReceiveMessageHandler {
                    override fun onReceiveMessage(message: ByteArray?) {
                        try {
                            val text = String(message!!, StandardCharsets.UTF_8)
                            writeToLog("PlayersOnlineActivity", "OnReceiveMessageHandler has received message: $text")
                            handleRabbitMQMessage(text)
                        } catch (e: Exception) {
                            writeToLog("PlayersOnlineActivity", "exception in onReceiveMessage: $e")
                        }
                    } // end onReceiveMessage
                }) // end setOnReceiveMessageHandler
                setContentView(R.layout.players_online_fragment) //this starts up the list view
            }
            setContext(this)
            writeToLog("PlayersOnlineActivity", "onCreate taskId: $taskId")
        }
    }

    private fun handleRabbitMQMessage(message: String) {
        if (mRabbitMQResponse.equals(message)) {
            writeToLog("PlayersOnlineActivity", "================> handleRabbitMQMessage() returning due to duplicate message: $message \n\n")
            return
        }
        mRabbitMQResponse = message
        if (message.startsWith("letsPlay")) {
            val playStringArray = message.split(",")
            if (playStringArray.size >= 3) {
                val opposingPlayerId = playStringArray[2]
                val opposingPlayerName = playStringArray[1]
                val i = Intent(mApplicationContext, GameActivity::class.java)

                val item = GameActivity.ParcelItems(123456789, "Shakespeare")
                i.putExtra(GameActivity.PARCELABLE_VALUES, item)

                i.putExtra(GameActivity.START_SERVER, "true")
                i.putExtra(GameActivity.PLAYER1_ID, mPlayer1Id)
                i.putExtra(GameActivity.PLAYER1_NAME, mPlayer1Name)
                i.putExtra(GameActivity.START_CLIENT_OPPONENT_ID, opposingPlayerId)
                i.putExtra(GameActivity.PLAYER2_NAME, opposingPlayerName)
                i.putExtra(GameActivity.START_FROM_PLAYER_LIST, "true")
                writeToLog("PlayersOnlineActivity", "starting client and server from new player: $opposingPlayerName")
                startActivity(appContext, i, null)
            }
            writeToLog("PlayersOnlineActivity", "got LetsPlay response received message: $message at: " +
                SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(Date())
            )
        }
        finish()
    }

    public override fun onPause() {
        super.onPause()
        writeToLog("PlayersOnlineActivity", "onPause called from Main Activity")
        if (mSelectedPosition == -1) {
            val urlData = "/gamePlayer/update/?id=$mPlayer1Id&onlineNow=false&opponentId=0&userName=$mPlayer1Name"
            val messageResponse = CoroutineScope(Dispatchers.Default).async {
                SendMessageToAppServer.main(urlData, mPlayersOnlineActivity as ToastMessage, mResources, false)
            }
            writeToLog("PlayersOnlineActivity", "onPause from Main Activity called to set onlineNow to false $messageResponse")
        }
        writeToLog("PlayersOnlineActivity", "onPause all done")
    }

    override fun onResume() { //only called when at least one opponent is online to select
        super.onResume()
        writeToLog("PlayersOnlineActivity", "onResume called from PlayersOnlineActivity")
    }

    override fun onStop() {
        super.onStop()
        writeToLog("PlayersOnlineActivity", "onStop called from main class")
    }

    override fun onDestroy() {
        super.onDestroy()
        writeToLog("PlayersOnlineActivity", "onDestroy called from main class")
    }

    /**
     * This is the "top-level" fragment, showing a list of items that the
     * user can pick.  Upon picking an item, it takes care of displaying the
     * data to the user as appropriate based on the current UI layout.
     */
    class PlayersOnlineFragment : ListFragment(), ToastMessage {
        private var mCurCheckPosition = 0
        private var mAlreadySelected = false

        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
            val view: View = inflater.inflate(R.layout.player_list_content, container, false)
            val adapter: ArrayAdapter<String> = ArrayAdapter<String>(
                inflater.context,
                android.R.layout.simple_list_item_1,
                mUserNames
            )
            listAdapter = adapter
            return view
        }

        override fun onResume() { //only called when at least one opponent is online to select
            super.onResume()
            writeToLog("PlayersOnlineFragment", "onResume called from PlayersOnlineFragment")
            mSelectedPosition = -1
        }

        override fun onPause() { // pause the PlayersOnlineFragment
            super.onPause()
            writeToLog("PlayersOnlineFragment", "onPause completed from PlayersOnlineFragment mSelectedPosition = $mSelectedPosition")
        }

        override fun onStop() {
            super.onStop()
            writeToLog("PlayersOnlineFragment", "onStop called from PlayersOnlineFragment")
        }

        override fun onDestroy() {
            super.onDestroy()
            writeToLog("PlayersOnlineFragment", "onDestroy called")
        }

        override fun onDestroyView() {
            super.onDestroyView()
            writeToLog("PlayersOnlineFragment", "onDestroyView() called")
        }

        override fun onDestroyOptionsMenu() {
            super.onDestroyOptionsMenu()
        }

        override fun onDetach() {
            super.onDetach()
        }

        override fun onSaveInstanceState(outState: Bundle) {
            super.onSaveInstanceState(outState)
            outState.putInt("curChoice", mCurCheckPosition)
        }

        override fun onListItemClick(l: ListView, v: View, position: Int, id: Long) {
            super.onListItemClick(l, v, position, id)
            writeToLog("PlayersOnlineActivity", ">>>>>>>>>>>> onListItemClick position: $position")
            if (mAlreadySelected) {
                // not sure if this is really needed
                // may want to reconsider making this class a distinct Rabbit MQ queue
                writeToLog("PlayersOnlineActivity", "<<<<<<<<<<<<<< onListItemClick already selected, I'm returning")
                //showCancelDialog()
                sendToastMessage("Please press the Back button to search for other cool players!")
                return
            }
            mAlreadySelected = true
            setUpClientAndServer(position)
            val qName = getConfigMap("RabbitMQQueuePrefix") + "-" + "playerList" + "-" + mUserIds[position]
            mRabbitMQConnection = setUpRabbitMQConnection(qName)
            mSelectedPosition = position
            //FIXME - clean up this code once fully debugged
            val rnds = (0..10000000).random() // generated random from 0 to 10,000,000 included
            val dateTime = SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(Date())
            writeToLog("PlayersOnlineActivity", "============> onListItemClick called  at: $dateTime")
            val messageToOpponent = "letsPlay,$mPlayer1Name,$mPlayer1Id,$rnds, $dateTime"
            GlobalScope.launch {
                    SendMessageToRabbitMQ().main(
                        mRabbitMQConnection,
                        qName,
                        messageToOpponent,
                        mPlayersOnlineActivity as ToastMessage,
                        mResources
                    )
            }
            if (mRabbitMQConnection != null) {
                closeRabbitMQConnection(mRabbitMQConnection!!)
                writeToLog("PlayersOnlineActivity", "closeRabbitMQConnection() all done")
            }
            writeToLog("PlayersOnlineActivity", "onListItemClick() completed selected opponent:  $position")
        }

        private fun showCancelDialog() {
            if (mCancelDialog == null) {
                mCancelDialog = createCancelDialog()
                mCancelDialog!!.show()
            } else {
                mCancelDialog!!.show()
            }
        }

        private fun createCancelDialog(): AlertDialog {
            if (mCancelDialog != null) {
                mCancelDialog!!.dismiss()
            }

            return AlertDialog.Builder(mApplicationContext!!)
                .setIcon(R.drawable.willy_shmo_small_icon)
                .setTitle("Trouble connecting?")
                .setMessage("Please press the Back button to search again")
                .setCancelable(false)
                .setNegativeButton("Cancel") { _, _ -> createCancelDialog() }
                .create()
        }

        private fun setUpRabbitMQConnection(qName: String): RabbitMQConnection {
            return runBlocking {
                withContext(CoroutineScope(Dispatchers.Default).coroutineContext) {
                    SetUpRabbitMQConnection().main(
                        qName,
                        mPlayersOnlineActivity as ToastMessage,
                        resources
                    )
                }
            }
        }

        private fun closeRabbitMQConnection(mRabbitMQConnection: RabbitMQConnection) {
            // terminate run loop in RabbitMQMessageConsumer
            writeToLog("PlayersOnlineActivity", "at start of closeRabbitMQConnection()")
            return runBlocking {
                writeToLog("PlayersOnlineActivity", "about to stop RabbitMQ consume thread")
                val messageToSelf = "finishConsuming,$mPlayer1Name,$mPlayer1Id"
                val myQName = getConfigMap("RabbitMQQueuePrefix") + "-" + "playerList" + "-" + mPlayer1Id
                withContext(CoroutineScope(Dispatchers.Default).coroutineContext) {
                    val sendMessageToRabbitMQ = SendMessageToRabbitMQ()
                    sendMessageToRabbitMQ.main(Companion.mRabbitMQConnection, myQName, messageToSelf, mPlayersOnlineActivity as ToastMessage, mResources)
                }
                writeToLog("PlayersOnlineActivity", "about to close RabbitMQ connection")
                withContext(CoroutineScope(Dispatchers.Default).coroutineContext) {
                    CloseRabbitMQConnection().main(mRabbitMQConnection, mPlayersOnlineActivity as ToastMessage, resources)
                }
                writeToLog("PlayersOnlineActivity", "about to Dispose RabbitMQ consumer")
                withContext(CoroutineScope(Dispatchers.Default).coroutineContext) {
                    val disposeRabbitMQTask = DisposeRabbitMQTask()
                    disposeRabbitMQTask.main(mMessageConsumer, mResources, mPlayersOnlineActivity as ToastMessage)
                }
            }
        }

        private fun setUpClientAndServer(which: Int) {
            val settings = mApplicationContext!!.getSharedPreferences(UserPreferences.PREFS_NAME, Context.MODE_PRIVATE)
            val editor = settings.edit()
            editor.putString("ga_opponent_screenName", mUserNames[which])
            editor.apply()

            val opponentUserId = mUserIds[which]
            CoroutineScope( Dispatchers.Default).launch {
                StartGameTask().main(mCallerActivity, mPlayer1Id!!, mPlayer1Name!!, opponentUserId!!, mUserNames[which]!!, resources)
            }
        }

        private fun showDetails(index: Int) {
            mCurCheckPosition = index
        }

        override fun sendToastMessage(message: String?) {
            val msg = SplashScreen.mErrorHandler!!.obtainMessage()
            msg.obj = message
            SplashScreen.mErrorHandler!!.sendMessage(msg)
        }
    }

    //we're creating a clone because removing an entry from the original TreeMap causes a problem for the iterator
    private val playersOnline: Unit
        get() {
            val users = parseUserList(mUsersOnline)
            val usersClone = users.clone() as TreeMap<*, *>
            //we're creating a clone because removing an entry from the original TreeMap causes a problem for the iterator
            val userKeySet: MutableSet<out Any> = usersClone.keys // this is where the keys (userNames) gets sorted
            val keySetIterator = userKeySet.iterator()
            while (keySetIterator.hasNext()) {
                val key = keySetIterator.next()
                val userValues = users[key]!!
                val userId = userValues["userId"]
                if (userId == (mPlayer1Id!!).toString()) //not going to play against myself on the network
                    users.remove(key)
            }
            val objectArray: Array<Any> = users.keys.toTypedArray()
            mUserNames = arrayOfNulls(objectArray.size)
            mUserIds = arrayOfNulls(objectArray.size)
            for (x in objectArray.indices) {
                mUserNames[x] = objectArray[x] as String
                val userValues = users[objectArray[x]]!!
                mUserIds[x] = userValues["userId"]
            }
        }

    private fun parseUserList(usersLoggedOn: String?): TreeMap<String, HashMap<String, String>> {
        val userTreeMap = TreeMap<String, HashMap<String, String>>()
        try {
            val jsonObject = JSONObject(usersLoggedOn)
            val userArray = jsonObject.getJSONArray("UserList")
            for (y in 0 until userArray.length()) {
                val userMapValues = HashMap<String, String>()
                val userValues = userArray.getJSONObject(y)
                val userId = userValues.getString("id")
                val userName = userValues.getString("userName")
                userMapValues["userId"] = userId
                userTreeMap[userName] = userMapValues
            }
        } catch (e: JSONException) {
            writeToLog("PlayersOnlineActivity", "parseUserList: " + e.message)
            sendToastMessage(e.message)
        }
        return userTreeMap
    }

    private val sharedPreferences: Unit
        get() {
            //TODO - Think about moving ga_users_online and ga_users_online_number to WillyShmoApplication
            val settings = getSharedPreferences(UserPreferences.PREFS_NAME, MODE_PRIVATE)
            mUsersOnline = settings.getString("ga_users_online", null)
            mPlayer1Id = settings.getInt(GameActivity.PLAYER1_ID, 0)
        }

    private fun setNumberOfPlayersOnline() {
        val settings = getSharedPreferences(UserPreferences.PREFS_NAME, Context.MODE_PRIVATE)
        val editor = settings.edit()
        editor.putInt("ga_users_online_number",  mUserNames.size)
        editor.apply()
    }

    /**
     * A simple utility Handler to display an error message as a Toast popup
     */
    class ErrorHandler : Handler(Looper.getMainLooper()) {
        override fun handleMessage(msg: Message) {
            Toast.makeText(mApplicationContext, msg.obj as String, Toast.LENGTH_LONG).show()
        }
    }

    override fun sendToastMessage(message: String?) {
        val msg = mErrorHandler!!.obtainMessage()
        msg.obj = message
        mErrorHandler!!.sendMessage(msg)
    }

    companion object {
        private var mSavedInstanceState: Bundle? = null
        private lateinit  var appContext: Context
        fun setContext(context: Context) {
            appContext = context
        }
        fun getContext(): Context {
            return appContext
        }
        private lateinit var mCallerActivity: PlayersOnlineActivity
        private lateinit var mUserNames: Array<String?>
        private lateinit var mUserIds: Array<String?>
        private var mPlayer1Id: Int? = null
        private var mPlayer1Name: String? = null
        private var mApplicationContext: Context? = null
        var mErrorHandler: ErrorHandler? = null
        private lateinit var mResources: Resources
        private var mPlayersOnlineActivity: PlayersOnlineActivity? = null
        private var mSelectedPosition = -1
        private var mRabbitMQConnection: RabbitMQConnection? = null
        private var mRabbitMQResponse: String? = null
        private var mMessageConsumer: RabbitMQMessageConsumer? = null
        private var mUsersOnline: String? = null
        private var mCancelDialog: AlertDialog? = null

        private fun writeToLog(filter: String, msg: String) {
            if ("true".equals(mResources.getString(R.string.debug), ignoreCase = true)) {
                Log.d(filter, msg)
            }
        }
    }
}